Add support for GSSAPI in Kafka scaler

novicr opened the pull request:
<!-- Thank you for contributing!

     Read more about how you can contribute in our contribution guide:
     https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md
-->

Added Kerberos authentication to Kafka scaler.  TriggerAuthentication can now accept additional parameters that will be passed to Sarama to authenticate scaler via Kerberos.

### Checklist

- [N/A] When introducing a new scaler, I agree with the [scaling governance policy](https://github.com/kedacore/governance/blob/main/SCALERS.md)
- [x] I have verified that my change is according to the [deprecations & breaking changes policy](https://github.com/kedacore/governance/blob/main/DEPRECATIONS.md)
- [x] Tests have been added
- [x] Changelog has been updated and is aligned with our [changelog requirements](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#Changelog)
- [N/A] A PR is opened to update our Helm chart ([repo](https://github.com/kedacore/charts)) *(if applicable, ie. when deployment manifests are modified)*
- [x] A PR is opened to update the documentation on ([repo](https://github.com/kedacore/keda-docs)) *(if applicable)*
- [x] Commits are signed with Developer Certificate of Origin (DCO - [learn more](https://github.com/kedacore/keda/blob/main/CONTRIBUTING.md#developer-certificate-of-origin-signing-your-work))

<!--
  Make sure to link the related issue for this change
  If it requires multiple PRs and/or a PR on another repo as well, please use "Relates to" instead.
-->
Fixes https://github.com/kedacore/keda/issues/4836

<!--
  Make sure to link the related PRs for changes such as documentation & Helm charts
-->
Relates to PR: https://github.com/kedacore/keda-docs/pull/1201

JorTurFer reviewed (commented):
Nice contribution!
Can we cover this new feature with an e2e tests somehow?

novicr commented:
My changes only consider the kafka scaler (based on sarama). The change is to collect relevant information in TriggerAuthentication and pass it along to sarama. Not sure about the e2e test - would require standing up kerberos infra as part of the test. I'd rather trust that sarama will do the right thing with inputs (it does). Unit tests ensure that correct values are passed along.

novicr commented on pkg/scalers/kafka_scaler.go:
```diff
@@ -51,6 +52,11 @@ type kafkaMetadata struct {
+	// GSSAPI
+	keytab        string
+	realm         string
+	krbConfigPath string
```
renamed the variable to `kerberosConfigPath`.  Note that variable here is different from the parameter name supplied in configuration.  Reason is, user supplies the configuration (ideally via ConfigMap) and internally Keda saves it into a file to pass the path to sarama.

JorTurFer commented on config/manager/manager.yaml:
```diff
@@ -77,13 +77,18 @@ spec:
             seccompProfile:
               type: RuntimeDefault
           volumeMounts:
+          - mountPath: /tmp
+            name: temp-vol
+            readOnly: false
```
we need to add this volume to the helm chart too

sansmoraxz replied:
Isn't it better to store the keytab in some folder other than tmp? I don't think it's a good practice to share tmp between pods.

JorTurFer replied:
What do you mean? We are not sharing `/tmp` between pods, each pod will create its own file each time that the scaler is built. But you have discovered me another potential issue, orphan files in `/tmp`. We need to delete the file on scaler closing

JorTurFer replied:
@novicr , could you add the code to delete the file on scaler close function? We should track if the file's been created and if it's been created, we should delete it on close to not keep orphan files. On long-running instances, the pod could have a lot of files (there are multiple reasons that could regenerate the scaler, generating new files)

sansmoraxz replied:
> We are not sharing `/tmp` between pods, each pod will create its own file each time that the scaler is built.

Ahh ok. My concern was that tmp generally implies that it's ephemeral.

It makes sense for the keytab to be in tmp then.

zroubalik replied:
I am not 100% sure we should store the keytab there, isn't it a possible security problem? Maybe an in-memory storage?

novicr replied:
> we need to add this volume to the helm chart too

Will do once we agree on the content.  There are clearly some legitimate changes pending here.

novicr replied:
> I am not 100% sure we should store the keytab there, isn't it a possible security problem? Maybe an in-memory storage?

Agreed.  Adding medium to keep this in memory.

novicr replied:
> @novicr , could you add the code to delete the file on scaler close function? We should track if the file's been created and if it's been created, we should delete it on close to not keep orphan files. On long-running instances, the pod could have a lot of files (there are multiple reasons that could regenerate the scaler, generating new files)

@JorTurFer , can scaler be closed and reopened in the context of the same Pod?  I did add file removal.

novicr replied:
> we need to add this volume to the helm chart too

Added a chart pr: https://github.com/kedacore/charts/pull/509

JorTurFer replied:
> @JorTurFer , can scaler be closed and reopened in the context of the same Pod? I did add file removal.

Yep. Ideally, the scaler is created once and closed only once because it's stored inside a scaler cache, but if there are failures, the scaler is regenerated. For example, if there is any transient error or any error in the upstream, the scaler is regenerated.
In the worst case with the upstream unreachable, the scaler will be opened and close twice per each reconcilliation loop, that's why I asked for removing the file on closing.

JorTurFer commented on pkg/scalers/kafka_scaler.go:
```diff
@@ -229,10 +215,105 @@ func parseKafkaAuthParams(config *ScalerConfig, meta *kafkaMetadata) error {
+	if config.AuthParams["kerberosConfig"] == "" {
+		return errors.New("no Kerberos configuration file (kerberosConfig) given")
+	}
+	path, err := saveToFile(config.AuthParams["kerberosConfig"])
```
Should we validate somehow the content before saving it? Could someone inject arbitrary code here?

JorTurFer commented on pkg/scalers/kafka_scaler.go:
```diff
@@ -398,6 +479,22 @@ func getKafkaClients(metadata kafkaMetadata) (sarama.Client, sarama.ClusterAdmin
+	if metadata.saslType == KafkaSASLTypeGSSAPI {
+		config.Net.SASL.Enable = true
+		config.Net.SASL.Mechanism = sarama.SASLTypeGSSAPI
+		config.Net.SASL.GSSAPI.ServiceName = "kafka"
```
Should this be a parameter?

novicr replied:
Don't believe this needs to be exposed to users - no value in it.  Scaler process will only need a single Kafka service.

sansmoraxz replied:
Needing to enable sasl can be mentioned in the docs. Giving the end user more control rather than some gotcha of it being automatically enabled.

novicr replied:
Nothing automatic is happening.  User specifies option - `sasl: gssapi`.  This way they declare intent to turn on sasl with kerberos auth.

JorTurFer reviewed (commented):
Is there any way to add an e2e tests which tests this?

zroubalik commented:
Also, how do we deal with the token refreshment? I am not Kerberors expert at all, though I think that it needs to be refreshed after some period of time.

novicr commented:
> Also, how do we deal with the token refreshment? I am not Kerberors expert at all, though I think that it needs to be refreshed after some period of time.

Sarama takes care of it.  We just provide the path to key tab

novicr commented:
I moved temp kerberos files into dedicated directory, just in case someone already mounts /tmp into their container.  Don't think we are likely to run into any conflicts with /tmp/kerberos, especially since kerberos isn't supported by keda yet

sansmoraxz commented:
once this merges. I will take a dig with `kafka-go`

novicr commented:
What are next steps on this?  There were couple small commits after approval. Need someone to approve again.  Also, how do we make sure this goes hand in hand with docs and charts PR's?

sansmoraxz commented:
E2E tests run by one of the core maintainers. Ignore my above approval, that was a misclick.

sansmoraxz commented:
They are still on vacation. So progress is slow.

JorTurFer commented:
> What are next steps on this? There were couple small commits after approval. Need someone to approve again. Also, how do we make sure this goes hand in hand with docs and charts PR's?

We check that all the PRs are ready to merge before merging it on any repo. It's a manual action that we do before merging.
I guess that during next days we will address all the reviews that we have because we have slowed the speed during summer vacations (personally, I had more than 100 notifications waiting review in GH xD)

novicr commented on config/manager/manager.yaml:
```diff
@@ -77,13 +77,19 @@ spec:
             seccompProfile:
               type: RuntimeDefault
           volumeMounts:
+          - mountPath: /tmp/kerberos
```
Note, the directory needs to be writable (by default manager pod runs with RO file system) and in-memory (to avoid exposing secrets on disk).

novicr replied:
I am not sure what you mean by "We can just mount an emptyDir on file system".  Can you please expand.  The code inside the pod is not able to create any directories.

zroubalik reviewed (commented):
I am still not sure we should add another mount point. If the file needs to be stored, can't we use some in-memory go filesystem implementation?

novicr commented:
> I am still not sure we should add another mount point. If the file needs to be stored, can't we use some in-memory go filesystem implementation?

Not sure what you mean, but open to any suggestions.  I can spin up in-mem fs in the code and create a file in there, but I don't see any way to pass that file to sarama.  Sarama accepts a parameter that if a file system path to the required files (kyetab and krb5.conf).  Any files created in my in-memory filesystem will not be visible to sarama.

zroubalik commented:
> > I am still not sure we should add another mount point. If the file needs to be stored, can't we use some in-memory go filesystem implementation?
> 
> Not sure what you mean, but open to any suggestions. I can spin up in-mem fs in the code and create a file in there, but I don't see any way to pass that file to sarama. Sarama accepts a parameter that if a file system path to the required files (kyetab and krb5.conf). Any files created in my in-memory filesystem will not be visible to sarama.

Right.

@sansmoraxz how does kafka-go client deal with kerberos auth? Does it need a file in a filesystem as well?

sansmoraxz commented:
Strictly speaking it's not something that's there in the official repo. Well there was a PR but not much progress was made.

The sasl interface should be loose enough to implement without issues. Have to look more in dept.

JorTurFer commented:
> * the existing Helm Chart supports adding arbitrary volume mounts, so it should be good from this side, CC @JorTurFer to confirm

yep! 
https://github.com/kedacore/charts/blob/274c330fa559e594d2eab1e897b581de6bc23e6d/keda/values.yaml#L475-L492

novicr commented:
Removed kerberos mount from config/manager.yaml
Added instructions to keda-docs to add dedicated mount when turning on kerberos auth (both via manager.yaml and charts).  Not sure my description in the docs is sufficient, please provide suggestions there.

When /tmp is not writable and no writable /tmp/kerberos is provided, error message now has instructions to follow documentation.

zroubalik commented on CHANGELOG.md:
```diff
@@ -77,7 +77,7 @@ New deprecation(s):
 
 - **General**: TODO ([#XXX](https://github.com/kedacore/keda/issues/XXX))
-
+- **Kafka**: Add support for Kerberos authentication (SASL / GSSAPI) ([#4836](https://github.com/kedacore/keda/issues/4836))
```
```suggestion
- **Kafka Scaler**: Add support for Kerberos authentication (SASL / GSSAPI) ([#4836](https://github.com/kedacore/keda/issues/4836))
```

Please, move this to Improvements section above. It is not a breaking change.

zroubalik commented:
/run-e2e kafka

zroubalik reviewed (approved):
LGTM, thanks for the contribution!
